src/mame/hp/hp9825.cpp

```diff
@@ -965,15 +891,215 @@ static INPUT_PORTS_START(hp9825)
+		m_printer_bitmap.plot_box(0 , m_printer_bitmap.height() - 1 , m_printer_bitmap.width() , 1 , PRINT_BG);
+	}
+
+	if ((m_printer_line > 0 && m_printer_line <= PRINT_EMPTY_TOP) ||
+		(m_printer_line > (PRINT_EMPTY_TOP + PRINT_MATRIX_H) && m_printer_line <= PRINT_CELL_H)) {
+		// Empty lines
+		for (unsigned i = 0; i < PRINT_PIXELS_ROW; i++) {
+			m_prt_graph_out->output(' ');
+		}
+	} else {
+		for (unsigned i = 0; i < PRINT_COLUMNS; i++) {
+			for (unsigned col = 0; col < PRINT_MATRIX_W; col++) {
+				uint8_t pixels = chargen[ m_printer_mem[ i ] & 0x7f ][ col ];
+				bool pixel = BIT(pixels , m_printer_line - PRINT_EMPTY_TOP - 1);
+				m_prt_graph_out->output(pixel ? '*' : ' ');
+				if (pixel) {
+					m_printer_bitmap.pix(m_printer_bitmap.height() - 1 , i * PRINT_CELL_W + col) = PRINT_FG;
+				}
+			}
+			m_prt_graph_out->output(' ');
+			m_prt_graph_out->output(' ');
+		}
+	}
+	m_prt_graph_out->output('\n');
+}
+
+uint32_t hp9825_state::printer_out_update(screen_device &screen, bitmap_rgb32 &bitmap, const rectangle &cliprect)
+{
+	s32 scroll = -s32((PRINT_WINDOWS - m_printer_window - 1) * PRINT_WIN_H);
+	copyscrollbitmap(bitmap , m_printer_bitmap , 0 , nullptr , 1 , &scroll , cliprect);
+
+	return 0;
+}
+
+static INPUT_PORTS_START(hp9825)
+	PORT_INCLUDE(hp98xx)
 
 	PORT_START("KEY_SCROLL")
 	PORT_BIT(IOP_MASK(0) , IP_ACTIVE_HIGH , IPT_KEYBOARD) PORT_NAME("Scroll up") PORT_CHANGED_MEMBER(DEVICE_SELF, hp9825_state, scroll_changed, 0)
 	PORT_BIT(IOP_MASK(1) , IP_ACTIVE_HIGH , IPT_KEYBOARD) PORT_NAME("Scroll down") PORT_CHANGED_MEMBER(DEVICE_SELF, hp9825_state, scroll_changed, 1)
 INPUT_PORTS_END
 
+// +---------------+
+// | hp9825a_state |
+// +---------------+
+class hp9825a_state : public hp9825_state
+{
+public:
+	hp9825a_state(const machine_config &mconfig, device_type type, const char *tag)
+		: hp9825_state(mconfig , type , tag)
+		, m_ram(*this, RAM_TAG)
+	{
+	}
+
+	void hp9825a(machine_config &config);
+
+protected:
+	virtual void machine_start() override;
+	virtual void device_reset() override;
+
+private:
+	required_device<ram_device> m_ram;
+
+	void cpu_mem_map(address_map &map);
+};
+
+void hp9825a_state::hp9825a(machine_config &config)
+{
+	hp9825_base(config);
+	m_cpu->set_addrmap(AS_PROGRAM , &hp9825a_state::cpu_mem_map);
+
+	RAM(config, m_ram);
+
+	m_ram->set_default_size("8K").set_extra_options("16K,24K,32K");
+
+	SOFTWARE_LIST(config, "optrom_list").set_original("hp9825a_rom");
+}
+
+void hp9825a_state::machine_start()
+{
+	hp9825_state::machine_start();
+
+	offs_t ram_start = 0x8000U - m_ram->size() / 2;
+
+	m_cpu->space(AS_PROGRAM).install_ram(ram_start, 0x7fff, m_ram->pointer());
+
+	for (auto& finder : m_rom_drawers) {
+		finder->set_rom_limit(ram_start);
+	}
+}
+
+void hp9825a_state::device_reset()
+{
+	hp98xx_state::device_reset();
+
+	auto space = &m_cpu->space(AS_PROGRAM);
+
+	for (auto& finder : m_rom_drawers) {
+		finder->install_rw_handlers(space , nullptr);
+	}
+}
```

Comment:
You should really be overriding `machine_reset` rather than `device_reset` for this kind of stuff in a `driver_device` implementation.  Also, how are you protecting against installing the same ROM devices multiple times on soft reset?